MSC3391: API to delete account data (#3391)
matrix-org/matrix-spec-proposals: proposals/3391-account-data-delete.md

Reviewer: @kerryarchibald, moving this to a thread.

> Does this mean there was never a working feature flag? I do seem to remember testing and seeing `DELETE` calls in the console, so possibly I tested it without the flag check, then added that later for completeness. I don't think the support check should be removed in js-sdk, will the flag be added to synapse?

There was not, no. it was missed in the original implementation. I've now added one in https://github.com/matrix-org/synapse/pull/15562.

Looking through the matrix-js-sdk client implementation, it looks like https://github.com/matrix-org/matrix-js-sdk/commit/b2a10e6 covers deleting _user_ account data, but [not room account data](https://github.com/matrix-org/matrix-js-sdk/blob/73ca9c9ed28847454e13da358c581769e695ff42/src/client.ts#L4217-L4224). Though I'm guessing there may not be a driving feature for for deleting room account data yet?

The mechanisms for deleting user vs room account data are exactly the same though, so I think it's safe to consider an implementation for only one of them sufficient.

> # MSC3391: Deleting account data

Author: At the moment I’m unsure if this MSC should move forward, considering some “account data as rooms” ideas I’ve seen floating around (disregarding the various problems that exist around that for the moment).

Would this MSC still be a good idea to move forward, considering it might interfere with future developments? Or could that be decided at review time?

{
  // ...
  "account_data": {
    "events": [
      {
        // ...
      }
    ],
    "removed_events": ["m.direct", "m.push_rules"] // <- NEW
  },
  // ...
  "rooms": {
    "join": {

      "!this-is-a-room:example.com": {
        // ...
        "account_data": {
          "events": [
            {
              // ...
            }
          ],
          "removed_events": ["m.tag", "m.fully_read"] // <- NEW
        }
        // ...
      },

    }
  },
  // ...
}

```

Providing an optional `removed_events` key per every `account_data` object,
containing an array which references the deleted account-data types.

These are the tags that were removed since `since` and `next_batch`, if `since` is specified and valid.

If between `since` and `next_batch` the account data has been deleted and re-created, this field shouldn't exist,
and data should be just put in `account_data.events` as if it's a normal change/creation.

If, for some reason, an event type exists in both `account_data.events` and `account_data.removed_events`, the reference in
`.removed_events` must be ignored.

Full-state syncs must not include `.removed_events`, but consequently clients must see anything

> in `events` as replacing what existed previously.

Author: I'm not 100% sure what the spec says about this on full sync, but this'd be a replace action on full sync hereafter, instead of a additive action to whatever internal state the client's cache would have.

## Potential issues

Desync is possible, and so if there's a situation where the client has a different view on account
data than the server, it should query the account data wherever possible.

> 

Reviewer 2: One issue is that a client that doesn't understand the `removed_events` property will still think that the data is still set.  One possible way to mitigate that would be to have the server also include it in `events`, but with `content` being an empty object, which is the closest to "deleting" that we currently have.  e.g.

```json5
{
  "account_data": {
    "events": [
      {
        "content": {},
        "type": "m.direct"
      }
    ],
    "removed_events": ["m.direct"]
  }
}

Author: I think that's a wart, but maybe it's a good approach while this is still unstable.

Should i add a mention of this to the unstable section? The intention is to remove this hack upon stabilisation, though.

Reviewer 3: If this is added, then there should also be a clarification saying that clients should remove `events` that are also in `removed_events`.

Author: I've added a note about this in [`88f7f7c` (#3391)](https://github.com/matrix-org/matrix-spec-proposals/pull/3391/commits/88f7f7cf824e6f6e39eed0359b7808dba875284b), would that resolve this thread?

Reviewer: I'd say so!

        }
        // ...
      },

    }
  },
  // ...
}

```

Providing an optional `removed_events` key per every `account_data` object,
containing an array which references the deleted account-data types.

These are the tags that were removed since `since` and `next_batch`, if `since` is specified and valid.

If between `since` and `next_batch` the account data has been deleted and re-created, this field shouldn't exist,
and data should be just put in `account_data.events` as if it's a normal change/creation.

If, for some reason, an event type exists in both `account_data.events` and `account_data.removed_events`, the reference in
`.removed_events` must be ignored.

Full-state syncs must not include `.removed_events`, but consequently clients must see anything
in `events` as replacing what existed previously.

#### Backwards Compatibility Note

Server implementations **may** include an empty (`{}`) `events` entry for deleted account data,
together with an entry in `removed_events`.

This is for backwards compatibility, as clients may or may not understand the `removed_events` key.

*The author thinks that this should only be necessary for a single matrix version,

> as clients adapt, but the SCT can - of course - have differing opinions.*

Reviewer: This should work, as long as no account data relies on a difference between a key being absent, and `key: {}`. Though given that account data could not be deleted before, I'd be skeptical of any implementation doing so.

Leaving in this optional backwards-compatibility hack, and then having another MSC to eventually remove the ability in a future spec version seems reasonable.

## Potential issues

Desync is possible, and so if there's a situation where the client has a different view on account
data than the server, it should query the account data wherever possible.

## Security considerations

No considerable security problems, other than the fact that a user can potentially delete important data.

Though, for the purposes of this proposal, that is seen as a proper feature of CRUD :)

## Unstable prefix

When implementing this proposal, servers should use the `org.matrix.msc3391` unstable prefix;

```
DELETE /_matrix/client/unstable/org.matrix.msc3391/user/{userId}/account_data/{type}

DELETE /_matrix/client/unstable/org.matrix.msc3391/user/{userId}/rooms/{roomId}/account_data/{type}
```

And `org.matrix.msc3391.removed_events` for `account_data` sync.

**Note:** As this operation would be largely "unknown" to clients,

> cache invalidation problems could occur as clients could aggressively cache account data.

Reviewer: The cache should be updated for a top-level account data key with the backwards-compatibility step mentioned above?

Author: Yeah, it seems that would purge old clients' caches, with new clients using the new method.

> # MSC3391: Deleting account data

Reviewer 4: > Synapse (#WIP)

@ShadowJonathan did you already have an implementation for this somewhere? I couldn't see anything in the Synapse PRs.

## Proposal

### Endpoints

> Account data and room data currently have the following endpoints;

Reviewer 5: please could you add a link to the relevant part of the existing spec.

### Sync

Account Data changes are announced through sync, this proposal also aims to add the following to sync;

```json5

{
  // ...
  "account_data": {
    "events": [
      {
        // ...
      }
    ],
    "removed_events": ["m.direct", "m.push_rules"] // <- NEW
  },
  // ...
  "rooms": {
    "join": {

      "!this-is-a-room:example.com": {
        // ...
        "account_data": {
          "events": [
            {
              // ...
            }
          ],
          "removed_events": ["m.tag", "m.fully_read"] // <- NEW
        }
        // ...
      },

    }
  },
  // ...
}

```

Providing an optional `removed_events` key per every `account_data` object, containing an array
which references the deleted account-data types.

These are the tags that were removed since `since` and `next_batch`, if `since` is specified and
valid.

If between `since` and `next_batch` the account data has been deleted and re-created, this field
shouldn't exist, and data should be just put in `account_data.events` as if it's a normal
change/creation.

If, for some reason, an event type exists in both `account_data.events` and

> `account_data.removed_events`, the reference in `.removed_events` must be ignored.

Author: I updated the document, this section no longer exists due to the changes.

Requests to `/sync` without a `since` token must not include `.removed_events`. Consequently,
anything present in `events` represents a full state of account data, and anything missing (from a
previous `since`-less sync) must be removed from the client's cache.

#### Backwards Compatibility Note

When a server announces it supports any version that does not include this MSC (via `/versions`), it
**must** include an empty (`{}`) `events` entry for deleted account data (per sync), together with

> the entry in `removed_events`.

Reviewer 5: what exactly would such an "empty" entry look like? you mean literally an empty object? but how would a client know what to do with that?

> the entry in `removed_events`.

Author: Yes, I meant an empty object, `{}`, here.

Current behaviour is that a client would replace their cached entry under that key with `{}` if they see that, effectively removing all data.

> the entry in `removed_events`.

Reviewer 5: > Current behaviour is that a client would replace their cached entry under that key with {} if they see that, effectively removing all data.

What is the "key" of an empty object?

(I think you might be assuming that account_data arrives over /sync as key/value pairs. That is not what the spec says, at least.)

> the entry in `removed_events`.

Author: I'm talking about both `/sync` and `/account_data`, and any other endpoint that might return account data.

> Why would it return {} rather than reverting to the same behaviour as before the account data was set in the first place (a 404, presumably)?

Because older clients might not know how to deal with a `404` when previously it'd return data, as that might be an assumption client/library authors make with how matrix is today; when account data is set, it is never un-set.

This MSC seeks to make it possible to un-set it, but older clients might crash and burn if presented with 404s here, and might not get the hint that data is removed with `removed_events`, and so continue on invalid assumptions.

(For example, if a `m.secret_storage.key` was present in one moment, and un-set in the next from a new client, the old client might perform undefined behaviour with the key(s) because it thinks that one key is still present.)

Thus, to gracefully introduce this new incompatible behaviour, the server should act as if `{}` is the new value for these deleted keys for compatibility reasons, until that's no longer necessary (when all versions that do not have account data removal are not supported anymore by that server software).

> the entry in `removed_events`.

Author: Sorry, yes, in my head I have been confusing this as a KV store, considering in all ways but mapping as representation, it acts like it.

Yes, I did mean `{}` for the `"content"` key for entries in `events`.

> the entry in `removed_events`.

Reviewer 5: Returning `{}` to `GET /_matrix/client/v3/user/{userId}/account_data/{type}` after it has been deleted seems... extreme.

Anyway, it's clear that this section needs clarifying before we can do any more useful review.

> the entry in `removed_events`.

Author: I don't think I will continue this MSC with that behaviour, as in [another thread](https://github.com/matrix-org/matrix-spec-proposals/pull/3391#discussion_r992616958) I have come to an agreement on how to proceed, which includes removing this exact `/account_data/` compatibility behaviour.

> the entry in `removed_events`.

Author: I updated the document, this section no longer exists due to the changes.

This is done for backwards compatibility reasons, as clients may or may not understand the
`removed_events` key when communicating with this server.

*This author believes that any matrix client will act responsibly when encountering a server

> supporting only matrix versions it does not understand, and fail-fast if this is the case.*

Author: @anoadragon453 / SCT, i'd like your opinion on this.

From my perspective, it seems reasonable to assume that a client will be tentative to interact with a server that only supports versions it does not "understand".

Especially in the context of this MSC, it'd be reasonable automatic deprecation of the backwards-compatible hack, if a client is expected to "know" at least one version the server announces, and all versions would include that MSC's behavior.

However, this is not a rule, and only an assumption, so I'm asking if this is a reasonable approach.

Reviewer: While reasonable indeed and a perfectly valid approach, it would come with the downside that even clients that support this MSC would still receive old entries in their account data. I don't expect homeservers to drop spec versions older than the one this MSC makes it into for... years most likely. Without that, I think this MSC loses its main beneficial property; the ability to remove keys from a user's account data.

Perhaps another approach could be to have clients signal to the homeserver that they support this feature. Something like adding an `exclude_deleted_account_data` boolean to `RoomEventFilter`? Alongside a query parameter to [`GET /_matrix/client/v3/user/{userId}/account_data/{type} `](https://spec.matrix.org/v1.3/client-server-api/#get_matrixclientv3useruseridaccount_datatype) and [`GET /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}`](https://spec.matrix.org/v1.3/client-server-api/#get_matrixclientv3useruseridroomsroomidaccount_datatype). It's not the cleanest solution and one we'd be stuck with for a long time... but perhaps arguably better than the alternative.

Open to thoughts from others on a better approach.

Author: First thoughts: I feel like a kickback mechanism like that would fall far beyond the scope of just this MSC, and feels superfluous to Matrix Versioning.

I'll have to see what I can do, I think it's better to look at the long term rather than short-term solutions for this.

Reviewer: Do any clients currently rely on account data entries with an empty object as a value? If not, could we just forgo the `removed_events` field entirely and just spec that an account data event with a content of `{}` means that the entry was deleted? I believe there's already some form of "`{}` means delete the account data entry is no longer in use" culture in clients today.

It would also mean that setting an entry with an empty content would no longer be valid.

Author: I don't know, then it wouldn't be a change that would require a spec change, but then it's more a semantic one regarding which apis do what (based on their names), and the fact that there's a guardrail on preventing setting `{}` on a key.

I don't know if that is worth the effort to spec and process, if a note declaring that "this is the intended way to delete a key" (without changing any functional behaviour) could be enough.

Author: Thank you for taking the time to discuss this, I agree with this approach, and will rewrite the MSC to take advantage of it.

However, I'd like to not leave the issue of "what is returned to devices who have seen the set account data key before" up to ambiguity.

So, I will write the above behaviour ("track which devices have seen it, and return `{}` to those that have") into the new prose, so that approach or a better one can be discussed.

Reviewer: @ShadowJonathan I think you may have misunderstood my last paragraph? Tracking which devices have seen the deletion go through (either via `/sync` or one of the `GET .../account_data/` endpoints) is only for the purpose of the homesever knowing when it can actually purge the account data from its database (to free up space). It's just a recommendation for space-saving in homeserver implementations; it doesn't change what is returned to clients.

* Clients that provide a `since` token before an account data entry deletion happened, should *always* receive `{}` content for the deleted event type in the `/sync` response.
* Clients that query a deleted account data entry via `GET /_matrix/client/v3/user/{userId}/account_data/{type}` or ` GET /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}` should *always* receive a `HTTP 404` / `M_NOT_FOUND` response.

Author: Thanks for clarifying, yes, I indeed didn't fully understand what you were going for, but now I think i do.

## Potential issues

Desync is possible, and so if there's a situation where the client has a different view on account

> data than the server, it should query the account data wherever possible.

Author: This was written when i was afraid that i'd miss to name some endpoints that are relevant for account data.

If not all endpoints are included into considering a new behaviour, the client could possibly get in a state where it uses removed account information as if it were live.

> data than the server, it should query the account data wherever possible.

Reviewer 5: ok, I don't find this section very clear. Perhaps you could update it with clarifications.

> data than the server, it should query the account data wherever possible.

Author: I updated the document, this section no longer exists due to the changes.

## Security considerations

No considerable security problems, other than the fact that a user can potentially delete important data.

Though, for the purposes of this proposal, that is seen as a proper feature of CRUD :)

## Unstable prefix

> When implementing this proposal, servers should use the `org.matrix.msc3391` unstable prefix;

Reviewer: Currently this MSC does not define any way for a client to know if it has been implemented (no `unstable_features` flag). That means that when a client tries to remove account data, there are 3 possible scenarios:

- The homeserver does not support this feature.
- The homesever supports an unstable version of this feature.
- (once the MSC is accepted) the homeserver supports a stable version of this feature.

It’d probably be a good idea to define `org.matrix.msc3391.unstable` and `org.matrix.msc3391.stable` `unstable_features` flags to help clients differentiate between all 3. (Also open to feedback on other solutions, as two unstable feature flags feels a bit messy.)

Author: I went with `org.matrix.msc3391` as an unstable key, i believe that clients can assume support for the unstable prefix if they see that key, and assume stable support if they observe the server `/versions` including a matrix version that contains this MSC.

# MSC3391: Deleting account data

## Proposal

### Endpoints

Account data and room data currently have the following endpoints;

```
GET /_matrix/client/r0/user/{userId}/account_data/{type}

PUT /_matrix/client/r0/user/{userId}/account_data/{type}

GET /_matrix/client/r0/user/{userId}/rooms/{roomId}/account_data/{type}

PUT /_matrix/client/r0/user/{userId}/rooms/{roomId}/account_data/{type}
```

This proposal aims to add the following two endpoints;

```
DELETE /_matrix/client/r0/user/{userId}/account_data/{type}

DELETE /_matrix/client/r0/user/{userId}/rooms/{roomId}/account_data/{type}
```

With no body.

### Sync

Account Data changes are announced through sync, this proposal also aims to add the following to sync;

```json5

{
  // ...

>   "account_data_removed": ["m.direct", "m.push_rules"], // <- NEW

Reviewer 6: Why not:

```
"account_data": {
  "events": [],
  "removed": []
}
```

Then the same schema would be extensible to state events (because of state res) too.

Author: ah heck, didnt see that, one second

the dedicated section)

### Endpoints

Account data and room data currently have the following endpoints:

Global Account Data,
[`GET`](https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv3useruseridaccount_datatype)
and
[`PUT`](https://spec.matrix.org/v1.4/client-server-api/#put_matrixclientv3useruseridaccount_datatype):
```
GET /_matrix/client/v3/user/{userId}/account_data/{type}

PUT /_matrix/client/v3/user/{userId}/account_data/{type}
```

And Room-specific Account Data,
[`GET`](https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv3useruseridroomsroomidaccount_datatype)
and
[`PUT`](https://spec.matrix.org/v1.4/client-server-api/#put_matrixclientv3useruseridroomsroomidaccount_datatype):
```
GET /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}

PUT /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}
```

This proposal aims to add the following two endpoints (with no request body):
```
DELETE /_matrix/client/v3/user/{userId}/account_data/{type}

DELETE /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}
```

> These, respectively, remove account-wide account data, and room-scoped account data.

Reviewer: Could you clarify what happens if you try to call either of these `DELETE` endpoints with a [room] account data type that doesn't exist?

I propose that `200 OK` / `{}` be returned, to ensure idempotent responses if called multiple times.

# MSC3391: Deleting account data

Currently, in Matrix, there is no semantic way to "remove" or "delete" account data. There is only a
method to create and update it, them being effectively the same operation, but no method to "remove"
it.

This violates the CRUD principle, and disallows "cleaning up" account data, or de-reference certain
keys used for certain actions, as then they will stay around forever.

## Proposal

This proposal aims to:
- Define "`{}`" account data content as effectively "deleted", alongside out-of-band deletion semantics.
- Allow `DELETE` HTTP methods on account data endpoints to perform the deletion.

The first, defining `{}` as "deleted", is to codify this defacto "nullification" into the spec as
expected behavior for deletion. This paves the way for better client interpretation of that data,
and would allow backwards compatibility when clients do `PUT` `{}` as account data. (More on this in
the dedicated section)

> ### Endpoints

Reviewer 7: I assume these endpoints would reject deleting `m.fully_read` like the current [setter endpoints](https://spec.matrix.org/v1.6/client-server-api/#server-behaviour-10)?

Author: Yes;

`For any account data key that cannot be set, the delete API will have a likewise error response.`

Would adding this satisfy this case?

Account data and room data currently have the following endpoints:

Global Account Data,
[`GET`](https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv3useruseridaccount_datatype)
and
[`PUT`](https://spec.matrix.org/v1.4/client-server-api/#put_matrixclientv3useruseridaccount_datatype):
```
GET /_matrix/client/v3/user/{userId}/account_data/{type}

PUT /_matrix/client/v3/user/{userId}/account_data/{type}
```

And Room-specific Account Data,
[`GET`](https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv3useruseridroomsroomidaccount_datatype)
and
[`PUT`](https://spec.matrix.org/v1.4/client-server-api/#put_matrixclientv3useruseridroomsroomidaccount_datatype):
```
GET /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}

PUT /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}
```

> This proposal aims to add the following two endpoints (with no request body):

Author: Note: Synapse is now also considering `PUT {}` as equivalent as `DELETE`, this is maybe worth considering to formalise.

```
DELETE /_matrix/client/v3/user/{userId}/account_data/{type}

DELETE /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}
```

These, respectively, remove account-wide account data, and room-scoped account data.

For idempotency reasons, these endpoints always return `200 OK`, with an empty JSON body `{}`.

#### Deleted account data responses

> Furthermore, when a client deletes account data, it must expect `/account_data/` to return a 404 on

Reviewer: ```suggestion
Furthermore, when a client deletes account data, it must expect the `GET` methods above to return a 404 on
```

for clarity

> the next request.

Reviewer: We ran into a Potential Issue:tm: when implementing this MSC in Synapse which I think would be good to document somewhere.

After an account data entry has been deleted, that change must be broadcast to all devices. Once that's been done, it would be ideal if we could remove the record of that account data type ever being set in the homeserver (for reduction of user data and (slight) storage benefits).

This gets a little complicated when you try to quantify at what point all devices have a user have seen the delete. One way to do this (what is [currently planned](https://github.com/matrix-org/synapse/issues/14244) for Synapse) is to track the devices that *haven't* yet seen the delete, and then tick them off the list once they have.

But what does it mean for a device to have *seen* an account data change? If they have `/sync`'d the change, we can likely say they've seen the change. We should be careful though, as the device may have error'd in processing the `/sync` response, and if it retries the request after we've deleted the row, then the response will not include the delete(!). Instead, we should only assume the client has successfully seen the delete when they make a *subsequent* `/sync` request with a `since` token that implies they've processed the delete (Synapse's `since`/`next_batch` tokens allow for this as they encode points in streams; other homeserver implementations may not have this property).

And what if a device requests `GET /_matrix/client/v3/user/{userId}/account_data/{type}` after a delete has occurred? Should we assume that the device has seen the delete? Initially I thought 'yes', but @reivilibre pointed out that this means a request to `GET /_matrix/client/v3/user/{userId}/account_data/{type}` could mean (if the device was the last to see the delete) that the delete would then not come down `/sync` - which feels like a massive footgun. As a result, we've decided not to assume a call to `GET /_matrix/client/v3/user/{userId}/account_data/{type}` means a device has seen a delete, and instead of use `/sync` requests.

---

All of this falls squarely into the area of implementation - but it may be useful to other homeserver implementations. None of this is an issue if the implementation chooses to keep "deleted" account data entries in their database forever.

Reviewer 8: > But what does it mean for a device to have _seen_ an account data change?

Presumably servers can implement the same logic as for when they need to figure out whether a device has seen a to-device message (so that the server can delete it)? Especially as to-device messages have a better case for robustness than account data.

Reviewer: Good point - it is indeed very similar. The only difference is that account data has this extra way of being accessed (via `GET /_matrix/client/v3/user/{userId}/account_data/{type}`). So whether you also count that as seeing the change is the only decision you need to make if you've already settled on a design for deleting old to-device messages.

Reviewer 5: > All of this falls squarely into the area of implementation

I'm not convinced it does, entirely. It seems important to me that we specify when exactly a client can expect the 'deleted' account data to be served.

Reviewer 9: I might be missing something obvious but how much is this problem different from any other change in account data? What if it's a good ol' PUT and some other device crashes at processing the /sync batch that has this change?

Reviewer 10: @ShadowJonathan this thread appears to be unresolved, and I think it's blocking FCP from starting.

Author: > As a result, we've decided not to assume a call to GET /_matrix/client/v3/user/{userId}/account_data/{type} means a device has seen a delete, and instead of use /sync requests.

Can we even tie access tokens to device IDs like this, or do we assume this per access token?

That brings up a whole can of worms for me too, and so i wish to assume that deletion of account data be a best-effort case only.

I think that enumerating and deleting account data only if all access tokens have seen it is not ideal from a performance perspective, since stale access tokens can "block" deletion of data like this, and it'd require a whole tracking system for what-access-token-has-seen-what in implementations, which i'm not aware exists right now.

---

Ultimately i'd wish to lean on @anoadragon453 for opinion on this; I think that `/account_data/` should reflect the current state at any time, and that it does not interact or touch "if the device has seen it or not", since any access token would already be syncing if it sends `account_data` requests like this, since a client is expected to sync continuously when its online.

However, what about clients that are *offline*? Do we expect implementations to internally store tombstones to their stream tokens, to notify remaining access tokens of such a delete? Won't that go against the storage savings, or will it increase performance as it only queries a "small" table every time such a sync happens, to search for such tombstones?

What if a client re-uses an old sync token, after it has restarted sync, or used an additional sync? This behaviour is undefined, and with the case of deleting data, even the reference of itself that it *has* been deleted, it might become problematic.

And i concur with @richvdh, this is not implementation-specific, as then different implementations will have different ways of specifying how and when data is deleted.

Imo, it should be tied to `/sync`, and `/account_data/` should always serve the current state, even if that results in a race condition between the two calls, as to which one returns the "this data has been deleted" state to the client first.

#### Backwards Compatibility

For backwards compatibility reasons, if a client `PUT`s `{}` for account data, that must be seen
equivalent as `DELETE`-ing that account data.

### Sync

Account Data changes are announced through sync, this proposal also aims to change this response slightly after account data deletion.

In `account_data.events` and `rooms.join.{room_id}.account_data.events`, a client must interpret
`{}` for an event content as it being deleted.

This must happen only on syncs with a `since` token.

## Security considerations

No considerable security problems, other than the fact that a user can potentially delete important data.

## Unstable prefix

When implementing this proposal, servers should use the `org.matrix.msc3391` unstable prefix:

```
DELETE /_matrix/client/unstable/org.matrix.msc3391/user/{userId}/account_data/{type}

DELETE /_matrix/client/unstable/org.matrix.msc3391/user/{userId}/rooms/{roomId}/account_data/{type}
```

> The `unstable_features` key for this MSC is `org.matrix.msc3391`.

Reviewer: Could you clarify this to say that, presumably, a boolean value of `true` means that MSC3391 is supported by the homeserver, and any other value or the key not being present means that support is not there?

Reviewer: @Johennes will Element clients (being as far as I'm aware the only clients that have implemented this MSC) want to eventually have a `org.matrix.msc3391.stable` key as well, to know that the feature is supported and that clients should make use of stabilised endpoints?

If so, we should add that to the MSC as well.

Author: > and any other value or the key not being present means that support is not there?

Ah, I thought this was implied, but i will clarify this 👍

Reviewer: > I _think_ so. @kerryarchibald is that correct?

Note that @kerryarchibald replied here: https://github.com/matrix-org/matrix-spec-proposals/pull/3391#issuecomment-1537535008

# MSC3391: Deleting account data

Currently, in Matrix, there is no semantic way to "remove" or "delete" account data. There is only a
method to create and update it, them being effectively the same operation, but no method to "remove"
it.

This violates the CRUD principle, and disallows "cleaning up" account data, or de-reference certain
keys used for certain actions, as then they will stay around forever.

## Proposal

This proposal aims to:
- Define "`{}`" account data content as effectively "deleted", alongside out-of-band deletion semantics.

> - Allow `DELETE` HTTP methods on account data endpoints to perform the deletion.

Reviewer 5: It's not entirely clear to me that we should have both of these mechanisms. My experience is that, where the protocol allows two ways of doing a thing, they will end up being subtly different somehow and it leads to compatibility problems (eg, one server implementation will forget to implement DELETE because the client they test with only uses PUT).

Given that we don't have any special server-side semantics for `PUT {}`, can we stick with that situation, and only do the special "deletion" behaviour for an actual `DELETE`?

Author: As an alternative, should we define `PUT {}` to return http 400, at the minuscule risk of breaking existing clients?

The first, defining `{}` as "deleted", is to codify this defacto "nullification" into the spec as
expected behavior for deletion. This paves the way for better client interpretation of that data,
and would allow backwards compatibility when clients do `PUT` `{}` as account data. (More on this in
the dedicated section)

### Endpoints

Account data and room data currently have the following endpoints:

Global Account Data,
[`GET`](https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv3useruseridaccount_datatype)
and
[`PUT`](https://spec.matrix.org/v1.4/client-server-api/#put_matrixclientv3useruseridaccount_datatype):
```
GET /_matrix/client/v3/user/{userId}/account_data/{type}

PUT /_matrix/client/v3/user/{userId}/account_data/{type}
```

And Room-specific Account Data,
[`GET`](https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv3useruseridroomsroomidaccount_datatype)
and
[`PUT`](https://spec.matrix.org/v1.4/client-server-api/#put_matrixclientv3useruseridroomsroomidaccount_datatype):
```
GET /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}

PUT /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}
```

> This proposal aims to add the following two endpoints (with no request body):

Reviewer 10: ```suggestion
This proposal adds the following two endpoints (with no request body):
```

MSCs are authoritative, generally.

```
DELETE /_matrix/client/v3/user/{userId}/account_data/{type}

DELETE /_matrix/client/v3/user/{userId}/rooms/{roomId}/account_data/{type}

> ```

Reviewer 10: it's implied, but not stated:
* Do these endpoints require authentication?
* Can these endpoints be rate limited?

Author: Yes and Yes, i'll add 👍

Author: Addressed in [`0653f34` (#3391)](https://github.com/matrix-org/matrix-spec-proposals/pull/3391/commits/0653f3468242c0fe18d90a16794c1488d30f54f2)

These, respectively, remove account-wide account data, and room-scoped account data.

For idempotency reasons, these endpoints always return `200 OK`, with an empty JSON body `{}`.

For any account data key that cannot be set (like `m.fully_read`), the delete API will have a likewise error response.

#### Deleted account data responses

Furthermore, when a client deletes account data, it must expect the `GET` methods above to return a 404 on
the next request.

#### Backwards Compatibility

For backwards compatibility reasons, if a client `PUT`s `{}` for account data, that must be seen
equivalent as `DELETE`-ing that account data.

### Sync

> Account Data changes are announced through sync, this proposal also aims to change this response slightly after account data deletion.

Reviewer 5: ```suggestion
Account Data changes are announced through sync; this proposal also aims to change this response slightly after account data deletion.
```

In `account_data.events` and `rooms.join.{room_id}.account_data.events`, a client must interpret

> `{}` for an event content as it being deleted.

Reviewer 10: Why the specific client behaviour? Clients are generally left to decide for themselves how they want to deal with this stuff. Some might choose to store it verbatim, others might not care. Some might even polyfill deleted/missing account data to ensure their code doesn't explode.

Reviewer 9: I guess the emphasis here should be on `{}` as the way to tell a client that the piece of account data is deleted, not on the client behaviour ("you shall delete this piece not store `{}`").

Author: This is also for backwards compatibility, to make sure this sudden approach does not break a lot of clients.

Author: This note also changes the *semantic* of `{}` to mean "deleted", partly for backwards compatibility, but also to harmonise other aspects of this proposal.

> This must happen only on syncs with a `since` token.

Reviewer 5: So to be clear: on syncs *without* a `since` token, clients must interpret `{}` as regular account data? Is that absolutely necessary? It seems like it's fiddly for a client to implement and it's not clear to me why it is so important.

Reviewer 9: I read this as an instruction to _server_ authors not to include deleted data (with `{}`) if `/sync` is called without `since`, even if the deletion has just occurred.

Reviewer 5: *much* clearer ;)

Author: Will improve the wording around thus 👍

Author: Addressed in [`0653f34` (#3391)](https://github.com/matrix-org/matrix-spec-proposals/pull/3391/commits/0653f3468242c0fe18d90a16794c1488d30f54f2)

> This must happen only on syncs with a `since` token.

Reviewer 5: I notice an earlier version of this MSC suggested a separate key for deleted account data (possibly to be used alongside returning `{}` in `events` for backwards compat). But I've not found where that idea was removed. Maybe you could discuss the reason not to use `deleted_events` in an "Alternatives" section?

Author: Will add 👍

Author: Addressed in [`0653f34` (#3391)](https://github.com/matrix-org/matrix-spec-proposals/pull/3391/commits/0653f3468242c0fe18d90a16794c1488d30f54f2)
